reasoner/resolution/resolver/RuleResolver.java

```diff
@@ -179,13 +179,15 @@ protected void initialiseDownstreamActors() {
     }
 
     @Override
-    protected ResponseProducer responseProducerCreate(Request request, int iteration) {
-        Traversal traversal = boundTraversal(rule.when().traversal(), request.partialAnswer().conceptMap());
-        ResourceIterator<ConceptMap> traversalIterator = traversalEngine.iterator(traversal).map(conceptMgr::conceptMap);
-        ResponseProducer responseProducer = new ResponseProducer(traversalIterator, iteration);
+    protected ResponseProducer responseProducerCreate(Request fromUpstream, int iteration) {
+        Traversal traversal = boundTraversal(rule.when().traversal(), fromUpstream.partialAnswer().conceptMap());
+        ResourceIterator<UpstreamVars.Derived> upstreamAnswers = traversalEngine.iterator(traversal).map(conceptMgr::conceptMap)
+                .flatMap(whenAnswer -> materialisations(fromUpstream, whenAnswer));
+        ResponseProducer responseProducer = new ResponseProducer(upstreamAnswers, iteration);
```

Comment:
note that a `Rule` uses `flatMap` instead of `.map()` because each answer can produce a iterator of answer materialisations now, instead of just one answer.